libpromises/evalfunction.c

```diff
@@ -509,67 +509,160 @@ static JsonElement* VarNameOrInlineToJson(EvalContext *ctx, const FnCall *fp, co
     return vardata;
 }
 
-static Rlist *GetHostsFromLastseenDB(Item *addresses, time_t horizon, bool return_address, bool return_recent)
+/*********************************************************************/
+
+static FnCallResult FnCallAnd(EvalContext *ctx,
+                              ARG_UNUSED const Policy *policy,
+                              ARG_UNUSED const FnCall *fp,
+                              const Rlist *finalargs)
+{
+
+    for (const Rlist *arg = finalargs; arg; arg = arg->next)
+    {
+        SyntaxTypeMatch err = CheckConstraintTypeMatch(fp->name, arg->val, CF_DATA_TYPE_STRING, "", 1);
+        if (err != SYNTAX_TYPE_MATCH_OK && err != SYNTAX_TYPE_MATCH_ERROR_UNEXPANDED)
+        {
+            FatalError(ctx, "Function '%s', %s", fp->name, SyntaxTypeMatchToString(err));
+        }
+    }
+
+    for (const Rlist *arg = finalargs; arg; arg = arg->next)
+    {
+        if (!IsDefinedClass(ctx, RlistScalarValue(arg)))
+        {
+            return FnReturnContext(false);
+        }
+    }
+
+    return FnReturnContext(true);
+}
+
+/************** FnCallHostsSeen utilities **************/
+
+typedef struct {
+    char *address;
+    char *hostkey;
+    char *lastseen;
+} HostData;
+
+static HostData *HostDataNew(const char *address, const char *hostkey, const char *buf)
+{
+    HostData *data = xmalloc(sizeof(HostData));
+    data->address = xstrdup(address);
+    data->hostkey = xstrdup(hostkey);
+    data->lastseen = xstrdup(buf);
+    return data;
+}
+
+static void HostDataFree(HostData *hd)
+{
+    assert(hd != NULL);
+    assert(hd->address != NULL
+           && hd->hostkey != NULL
+           && hd->lastseen != NULL);
+
+    free(hd->address);
+    free(hd->hostkey);
+    free(hd->lastseen);
+    free(hd);
+}
+
+typedef enum {
+    NAME,
+    ADDRESS,
+    KEY,
+    NONE
+} HostsSeenFormatOption;
```

Comment:
```suggestion
} HostsSeenFieldOption;
```

https://docs.cfengine.com/docs/3.15/reference-functions-hostsseen.html